docs: add comparison of WSL versions and their impact on Ubuntu (#1307)
canonical/ubuntu-pro-for-wsl: docs/explanation/compare-wsl-versions.md

---
myst:
  html_meta:
    "description lang=en": "Read about the differences between major versions of WSL, such as WSL 1 and WSL 2, and how it affects Ubuntu on WSL."
---

# Comparing WSL versions for Ubuntu on WSL

> This page goes over the main differences between different major versions of WSL and how they affect Ubuntu on WSL.

Reviewer: ```suggestion
This page explains the main differences between the major versions of WSL and how they affect Ubuntu on WSL.
```

> ## Background

Reviewer: ```suggestion
## Background on WSL versions
```

Just making the header more explicit.

> WSL has had two major versions: WSL 1 and WSL 2. Due to the significant architectural differences between the two versions, the behaviour and experience of Ubuntu on WSL differ
> quite a bit when using one or the other. A single Windows install can have multiple instances of Ubuntu, some using WSL 1 and some with WSL 2, but the default is WSL 2.

Reviewer: ```suggestion
WSL has had two major versions: WSL 1 and WSL 2. The default version on Windows is WSL 2, although it is possible to run individual Ubuntu instances using either version.

There are significant architectural differences between the two versions, which can affect the behaviour of Ubuntu on WSL.
```

Some changes here for simplicity and flow:

* Start with the simple case (the default version) before elaborating the complex case (multiple instances on different versions)
* Rearrange sentence to avoid excess verbiage ("quite a bit", "one or the other", "and experience")
* End section with a sentence that leads most naturally to the next section (architectural differences affecting behaviour)

> ### Comparing WSL 1 and WSL 2

Reviewer: ```suggestion
## Overview of major differences between WSL 1 and WSL 2
```

* I think this section stands alone and does not need to be nested under background
* I suggest giving it a more explicit heading and one that is distinct from the page name

> The primary difference between WSL 1 and WSL 2 is that WSL 1 acts like a compatibility layer for Linux binaries with partial support for Linux system calls, while WSL 2 utilises a

Reviewer: ```suggestion
The primary difference between WSL 1 and WSL 2 is that WSL 1 acts like a compatibility layer for Linux binaries, with partial support for Linux system calls, while WSL 2 utilises a
```

This is a somewhat complex sentence, so it's good to break it up with punctuation where appropriate.

> complete Linux kernel inside a managed VM. This has a number of consequences in terms of capabilities and performance.

Reviewer: ```suggestion
complete Linux kernel inside a managed VM, with more complete support for Linux system calls. This difference has a number of consequences for the capabilities and performance of WSL when running Ubuntu.
```

Suggestions here primarily aim to make the text more explicit and to keep the main thread running (WSL, Ubuntu).

As WSL 1 only has partial support for Linux system calls, some programs may not behave as expected when using WSL 1. Additionally, WSL 1 lacks support for graphical applications and systemd, a system and service management suite.

Generally, applications that are entirely contained in WSL will be faster on WSL 2, particularly for file IO intensive applications. However, WSL 1 has faster access to files mounted from Windows, so for certain use cases, WSL 1 may be faster.

> ## What this means

Reviewer: ```suggestion
## Summary of feature support across WSL versions
```

The differences between WSL 1 and WSL 2 and how it affects Ubuntu can be summarised with the table below:

| Feature                                                               |         WSL 1          |       WSL 2        |
| --------------------------------------------------------------------- | :--------------------: | :----------------: |
| Integration between Windows and Linux                                 |   {bdg-success}`Yes`   | {bdg-success}`Yes` |
| Fast boot times                                                       |   {bdg-success}`Yes`   | {bdg-success}`Yes` |
| Small resource foot print compared to traditional Virtual Machines    |   {bdg-success}`Yes`   | {bdg-success}`Yes` |
| Managed VM                                                            |    {bdg-danger}`No`    | {bdg-success}`Yes` |
| Full Linux Kernel                                                     |    {bdg-danger}`No`    | {bdg-success}`Yes` |
| Full system call compatibility                                        |    {bdg-danger}`No`    | {bdg-success}`Yes` |

> | Performance across OS file systems                                    |   {bdg-success}`Yes`   |  {bdg-danger}`No`  |

Reviewer: ```suggestion
| High performance across OS file systems                                    |   {bdg-success}`Yes`   |  {bdg-danger}`No`  |
```

Presumably, there is some performance for WSL 2 here, it's just not high relative to WSL 1.

| systemd support                                                       |    {bdg-danger}`No`    | {bdg-success}`Yes` |
| IPv6 support                                                          |   {bdg-success}`Yes`   | {bdg-success}`Yes` |
| Graphical application support                                         |    {bdg-danger}`No`    | {bdg-success}`Yes` |
| [Snap](https://snapcraft.io/) support                                 |    {bdg-danger}`No`    | {bdg-success}`Yes` |
| [cloud-init](https://cloud-init.io/) support                          |    {bdg-danger}`No`    | {bdg-success}`Yes` |
| [Ubuntu Pro for WSL](../tutorials/getting-started-with-up4w/) support |    {bdg-danger}`No`    | {bdg-success}`Yes` |
| [Landscape](ref::landscape-client) support                            |    {bdg-danger}`No`    | {bdg-success}`Yes` |
| [Ubuntu Pro Client](ref::ubuntu-pro-client) support                   | {bdg-warning}`Partial` | {bdg-success}`Yes` |

### How this affects Ubuntu for WSL

> Beyond the lack of support for graphical applications on WSL 1, the differences between WSL 1 and WSL 2 have notable consequences for Ubuntu for WSL. Most of this stems from the lack of systemd support on WSL 1. Systemd is a system and service management suite that Ubuntu and many applications for Ubuntu rely on. Most notably, [Snaps](https://snapcraft.io/) and [cloud-init](https://cloud-init.io/) do not work on WSL 1 due to the lack of support for systemd.

Reviewer: ```suggestion
The differences between WSL 1 and WSL 2 have notable consequences for Ubuntu on WSL, largely due to the lack of systemd support on WSL 1. Systemd is a system and service management suite that Ubuntu and many applications for Ubuntu depend on. Most notably, [Snaps](https://snapcraft.io/) and [cloud-init](https://cloud-init.io/) do not work on WSL 1 due to the lack of support for systemd. There is also a lack of support for graphical applications on WSL 1.
```

Slight modifications to emphasise the main point being made about systemd support. Some minor rewording also.

> While Ubuntu for WSL will still work on WSL 1, the experience may be degraded. Thus, we generally recommend sticking to WSL 2 unless your use case has specific requirements that make WSL 1 a better fit. For instance, if you make extensive use of file interoperability between WSL and Windows, you may experience better performance with WSL 1.

Reviewer: ```suggestion
While Ubuntu works on WSL 1, the experience may be degraded relative to WSL 2. Thus, we generally recommend using WSL 2 unless you have specific requirements. For example, if you make extensive use of file interoperability between WSL and Windows, you may benefit from the better performance of WSL 1 in that specific area.
```

Minor changes for simplicity and clarity.

### How this affects Ubuntu Pro for WSL

> [Ubuntu Pro for WSL](../tutorials/getting-started-with-up4w/) relies on systemd for much of its functionality, so it does not support WSL 1. Additionally, [Landscape](ref::landscape-client) is also not supported on WSL 1 for the same reason. You can still attach your WSL 1 instance to [Ubuntu Pro](https://documentation.ubuntu.com/pro/) using the [Ubuntu Pro Client](ref::ubuntu-pro-client), but some of the systems that come with Ubuntu Pro may not work on WSL 1.

Reviewer: ```suggestion
[Ubuntu Pro for WSL](../tutorials/getting-started-with-up4w/) relies on systemd for much of its functionality, including automatic Pro attachment, so it does not support WSL 1. For the same reason, the [Landscape](ref::landscape-client) tool for remote management of Ubuntu instances is also not supported on WSL 1. You can still manually attach your WSL 1 instance to [Ubuntu Pro](https://documentation.ubuntu.com/pro/) using the [Ubuntu Pro Client](ref::ubuntu-pro-client), although some of Ubuntu Pro's features may not work on WSL 1.
```

Some changes to make concrete what we are talking about when we say Ubuntu Pro for WSL, versus Landscape, versus Ubuntu Pro client.
